### PR TITLE
Adding in retry for 503 rate limiting errors.

### PR DESCRIPTION
**Pull request by adam-in-ict**

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -87,6 +99,35 @@ export class APIClient {
       body,
     });
 
+    // According to BambooHR documentation, a 429 error is a limit exceeded
+    // regarding adding employees.  A 503 is a "currently unavailable" 
+    // error that is most commonly due to rate limiting.  It may contain
+    // a 'Retry-After' value in the header specifying how long to wait.
+    // Unfortunately, we don't have documenation or examples for this, and
```
Comment (VDubber) [inline src/client.ts:106]: nit: documentation

Comment (adam-in-ict) [reply src/client.ts:106]: Went ahead and updated my inline comment and added information in documentaiton.md after I saw this. :)

Comment (VDubber) [inline src/client.ts:106]: I have an old friend working at BambooHR... Wonder if he'd know... 😏

File: src/client.ts [hunk 2]
```diff
@@ -87,6 +99,35 @@ export class APIClient {
...
+    // the HTTP spec states ths can either be a second value or an HTTP-date.
+    // For now, we're checking if the value is a number and attempting to 
+    // handle either option with a max upper limit for how long we'll wait
+    // to avoid worst case scenarios for failed parsing.
+    if(response.status === 503) {
+      let secondsToAwait = DEFAULT_RATE_LIMIT_WAIT * attemptCounter;
+      if(response.headers['Retry-After']) {
```
Comment (VDubber) [inline src/client.ts:114]: nit: consider pulling `response.headers['Retry-After']` out into a variable.

File: src/client.ts [hunk 3]
```diff
@@ -87,6 +99,35 @@ export class APIClient {
...
+    // According to BambooHR documentation, a 429 error is a limit exceeded
+    // regarding adding employees.  A 503 is a "currently unavailable" 
+    // error that is most commonly due to rate limiting.  It may contain
+    // a 'Retry-After' value in the header specifying how long to wait.
+    // Unfortunately, we don't have documenation or examples for this, and
+    // the HTTP spec states ths can either be a second value or an HTTP-date.
+    // For now, we're checking if the value is a number and attempting to 
+    // handle either option with a max upper limit for how long we'll wait
+    // to avoid worst case scenarios for failed parsing.
+    if(response.status === 503) {
```
Comment (VDubber) [inline src/client.ts:112]: nit: consider pulling out this block of code into a separate function for readability.

File: src/client.ts [hunk 4]
```diff
@@ -87,6 +99,35 @@ export class APIClient {
...
+      let secondsToAwait = DEFAULT_RATE_LIMIT_WAIT * attemptCounter;
+      if(response.headers['Retry-After']) {
+        this.logger.info(`Received a 503 response with a Retry-After value of `, response.headers['Retry-After']);
+        if (typeof response.headers['Retry-After'] === 'number') {
+          secondsToAwait = Math.min(MAX_RATE_LIMIT_WAIT, response.headers['Retry-After']);
+        }
+        const currentDateTime = new Date(Date.now());
```
Comment (VDubber) [inline src/client.ts:119]: It seems like lines 118 - 120 should be wrapped in an `else` block or else the `secondsToAwait` will be overridden every time by the attempt to use Retry-After as a date.

Comment (adam-in-ict) [reply src/client.ts:119]: Whoops!  Thanks!

File: src/client.test.ts [hunk 5]
```diff
@@ -37,7 +40,7 @@ describe('client APIs', () => {
       },
     });
 
-    const client = new APIClient(integrationConfig);
+    const client = new APIClient(integrationConfig, createMockIntegrationLogger());
```
Comment (VDubber) [inline src/client.test.ts:43]: Could we get some test coverage around the new code block? 
If it's not easy, probably not worth it.

Comment (adam-in-ict) [reply src/client.test.ts:43]: I've kicked around a few ideas, but wasn't terribly happy with anything yet.  Certainly open to further feedback.  I thought about a simple test to make sure it waited the proper number of seconds, but TBH I've never performed that kind of test to make sure it awaited the proper amount of milliseconds and wasn't even sure that would be worth the effort.

File: src/client.ts (hunk 4 above)
Comment (adam-in-ict) [reply src/client.ts]: Updated in the latest commit.

File: src/client.ts (hunk 3 above)
Comment (adam-in-ict) [reply src/client.ts:102]: Updated in the latest commit.

File: src/client.ts (hunk 2 above)
Comment (adam-in-ict) [reply src/client.ts]: Updated in the latest commit.

**Review by VDubber — APPROVED** [review]
Looks great 👍

**Pull request merged**